[feat-record-getInfos] 가계부 메인화면 지출 조회 기능 구현

71mC40H0 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 조회시작일과 조회종료일, 카테고리들(Optional)을 받아 사용일자의 내림차순으로 Return
- 카테고리가 null 또는 empty인 경우 모든 범위 카테고리 지정
- 동적 쿼리 작성을 위한 Querydsl 설정
- List<Category> 검증을 위한 Custom annotation 및 validator 구현
- 프론트에서 요청하는 날짜의 양식을 LocalDate 타입로 변환하기 위한 Converter와 Util 클래스 작성
- Converter의 경우 WebConfig에 등록하여 @ReqestParam과 @PathVariable에 적용

**TO-BE**
- 지출 분석 화면 조회 기능 구현
- Custom Exception 추가 및 코드 리팩토링 

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [x] API 테스트
- API 테스트는 PostMan으로 진행

**지출 등록**
memberId: 1로 요청
|항목|요청|기대결과|실제결과|비고|
|---|---|---|---|---|
|성공|{"amount" : 10000,"category" : "FOOD",<br/>"paidFor" : "가게",<br/>"memo" : "메모",<br/>"useDate" : "2/29/2020",<br/>"payType" : "CASH"}|Create Success|Create Success|DB에서 저장 확인|
|실패 - 잘못된 amount|{"amount" : 0,<br/>"category" : "FOOD",<br/>"paidFor" : "가게",<br/>"memo" : "메모",<br/>"useDate" : "2/29/2020",<br/>"payType" : "CASH"}|{"amount": "1 이상이어야 합니다"}|{"amount": "1 이상이어야 합니다"}| |
|실패 - 잘못된 category|{"amount" : 10000,<br/>"category" : "FOODS",<br/>"paidFor" : "가게",<br/>"memo" : "메모",<br/>"useDate" : "2/29/2020",<br/>"payType" : "CASH"}|{"category": "Enum에 없는 값입니다"}|{"category": "Enum에 없는 값입니다"}| |
|실패 - 잘못된 useDate 패턴|{"amount" : 10000,<br/>"category" : "FOOD",<br/>"paidFor" : "가게",<br/>"memo" : "메모",<br/>"useDate" : "2/290/2020",<br/>"payType" : "CASH"}|{"useDate": "유효하지 않은 날짜 형식입니다"}|{"useDate": "유효하지 않은 날짜 형식입니다"}| |
|실패 - 유효하지 않은 날짜 형식|{"amount" : 10000,<br/>"category" : "FOOD",<br/>"paidFor" : "가게",<br/>"memo" : "메모",<br/>"useDate" : "2/29/2023",<br/>"payType" : "CASH"}|{"message": "유효하지 않은 날짜 형식입니다: 2/29/2023"}|{"message": "유효하지 않은 날짜 형식입니다: 2/29/2023"}|패턴은 맞으나 유효한 날짜값이 아닌 경우|
|실패 - 잘못된 payType|{"amount" : 10000,<br/>"category" : "FOOD",<br/>"paidFor" : "가게",<br/>"memo" : "메모",<br/>"useDate" : "2/29/2020",<br/>"payType" : "CASHS"}|{"payType": "Enum에 없는 값입니다"}|{"payType": "Enum에 없는 값입니다"}| |
|실패 - 필수 입력값 미입력|{"amount" : "",<br/>"category" : "",<br/>"paidFor" : "",<br/>"memo" : "",<br/>"useDate" : "",<br/>"payType" : ""}|{"amount": "1 이상이어야 합니다",<br/>"payType": "Enum에 없는 값입니다",<br/>"useDate": "유효하지 않은 날짜 형식입니다",<br/>"paidFor": "공백일 수 없습니다",<br/>"category": "Enum에 없는 값입니다"}|{"amount": "1 이상이어야 합니다",<br/>"payType": "Enum에 없는 값입니다",<br/>"useDate": "유효하지 않은 날짜 형식입니다",<br/>"paidFor": "공백일 수 없습니다",<br/>"category": "Enum에 없는 값입니다"}|memo는 필수값 아님<br/>응답순서는 바뀔 수 있음|

**지출 수정**
recordId: 1, memberId: 1인 Record 저장되어 있는 상태
|항목|요청|기대결과|실제결과|비고|
|---|---|---|---|---|
|성공|memberId: 1<br/>recordId: 1<br/>{"amount" : 50000,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/28/2023",<br/>"payType" : "CARD"}|Update Success|Update Success|DB에서 수정 확인|
|실패 - 존재하지 않는 recordId|memberId: 1<br/>recordId: 2<br/>{"amount" : 50000,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/28/2023",<br/>"payType" : "CARD"}|콘솔창:<br/>java.lang.RuntimeException: Not found record|콘솔창:<br/>java.lang.RuntimeException: Not found record|Custom Exception과 연동 전으로 임시로 RuntimeException 발생<br/>Custom Exception으로 응답 수정 예정|
|실패 - 일치하지 않는 memberId|memberId: 2<br/>recordId: 1<br/>{"amount" : 50000,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/28/2023",<br/>"payType" : "CARD"}|콘솔창:<br/>java.lang.RuntimeException: Not match member|콘솔창:<br/>java.lang.RuntimeException: Not match member|Custom Exception과 연동 전으로 임시로 RuntimeException 발생<br/>Custom Exception으로 응답 수정 예정|
|실패 - 잘못된 amount|memberId: 1<br/>recordId: 1<br/>{"amount" : 0,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/28/2023",<br/>"payType" : "CARD"}|{"amount": "1 이상이어야 합니다"}|{"amount": "1 이상이어야 합니다"}| |
|실패 - 잘못된 category|memberId: 1<br/>recordId: 1<br/>{"amount" : 50000,<br/>"category" : "PETS",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/28/2023",<br/>"payType" : "CARD"}|{"category": "Enum에 없는 값입니다"}|{"category": "Enum에 없는 값입니다"}| |
|실패 - 잘못된 useDate 패턴|memberId: 1<br/>recordId: 1<br/>{"amount" : 50000,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/280/2023",<br/>"payType" : "CARD"}|{"useDate": "유효하지 않은 날짜 형식입니다"}|{"useDate": "유효하지 않은 날짜 형식입니다"}| |
|실패 - 유효하지 않은 날짜 형식|memberId: 1<br/>recordId: 1<br/>{"amount" : 50000,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/29/2023",<br/>"payType" : "CARD"}|{"message": "유효하지 않은 날짜 형식입니다: 2/29/2023"}|{"message": "유효하지 않은 날짜 형식입니다: 2/29/2023"}|패턴은 맞으나 유효한 날짜값이 아닌 경우|
|실패 - 잘못된 payType|memberId: 1<br/>recordId: 1<br/>{"amount" : 50000,<br/>"category" : "PET",<br/>"paidFor" : "동물병원",<br/>"memo" : "진료",<br/>"useDate" : "2/28/2023",<br/>"payType" : "CARDS"}|{"payType": "Enum에 없는 값입니다"}|{"payType": "Enum에 없는 값입니다"}| |
|실패 - 필수 입력값 미입력|memberId: 1<br/>recordId: 1<br/>{"amount" : "",<br/>"category" : "",<br/>"paidFor" : "",<br/>"memo" : "",<br/>"useDate" : "",<br/>"payType" : ""}|{"amount": "1 이상이어야 합니다",<br/>"payType": "Enum에 없는 값입니다",<br/>"useDate": "유효하지 않은 날짜 형식입니다",<br/>"paidFor": "공백일 수 없습니다",<br/>"category": "Enum에 없는 값입니다"}|{"amount": "1 이상이어야 합니다",<br/>"payType": "Enum에 없는 값입니다",<br/>"useDate": "유효하지 않은 날짜 형식입니다",<br/>"paidFor": "공백일 수 없습니다",<br/>"category": "Enum에 없는 값입니다"}|memo는 필수값 아님<br/>응답순서는 바뀔 수 있음|

**지출 삭제**
recordId: 1, memberId: 1인 Record 저장되어 있는 상태
|항목|요청|기대결과|실제결과|비고|
|---|---|---|---|---|
|성공|recordId: 1, memberId: 1|Delete Success|Delete Success|DB에서 삭제 확인|
|실패 - 존재하지 않는 recordId|recordId: 2, memberId: 1|콘솔창:<br/>java.lang.RuntimeException: Not found record|콘솔창:<br/>java.lang.RuntimeException: Not match member|Custom Exception과 연동 전으로 임시로 RuntimeException 발생<br/>Custom Exception으로 응답 수정 예정|
|실패 - 일치하지 않는 memberId|recordId: 1, memberId: 2|콘솔창:<br/>java.lang.RuntimeException: Not match member|콘솔창:<br/>java.lang.RuntimeException: Not match member|Custom Exception과 연동 전으로 임시로 RuntimeException 발생<br/>Custom Exception으로 응답 수정 예정|

**지출 조회(가계부 메인)**
memberId: 1

***Record Table***
|id|memberId|amount|category|paidFor|memo|useDate|payType|
|---|---|---|---|---|---|---|---|
|1|1|100000|FOOD|식당|고기집|2023-06-06|CARD|
|2|1|200000|FOOD|식당|고기집|2023-06-07|CARD|
|3|2|50000|PET|동물병원|강아지진료|2023-06-02|CARD|
|4|1|100000|CLOTHES|백화점|옷쇼핑|2023-05-05|CARD|
|5|1|20000|BEAUTY|미용실|커트|2023-06-01|CASH|
|6|1|10000|FOOD|국밥집|국밥|2023-06-02|CARD|
|7|1|1200|FOOD|편의점|껌|2023-06-01|CASH|


|항목|요청|기대결과|실제결과|비고|
|---|---|---|---|---|
|성공 - 필수값만 입력|memberId: 1<br/>startDate: 6/1/2023<br/>endDate: 6/2/2023|[{"recordId": 6,<br/>"amount": 10000,<br/>"category": "식비",<br/>"paidFor": "국밥집",<br/>"memo": "국밥",<br/>"useDate": "2023-06-02",<br/>"payType": "카드"},<br/>{"recordId": 5,<br/>"amount": 20000,<br/>"category": "뷰티",<br/>"paidFor": "미용실",<br/>"memo": "커트",<br/>"useDate": "2023-06-01",<br/>"payType": "현금" },<br/>{"recordId": 7,<br/>"amount": 1200,<br/>"category": "식비",<br/>"paidFor": "편의점",<br/>"memo": "껌",<br/>"useDate": "2023-06-01",<br/>"payType": "현금"}]|[{"recordId": 6,<br/>"amount": 10000,<br/>"category": "식비",<br/>"paidFor": "국밥집",<br/>"memo": "국밥",<br/>"useDate": "2023-06-02",<br/>"payType": "카드"},<br/>{"recordId": 5,<br/>"amount": 20000,<br/>"category": "뷰티",<br/>"paidFor": "미용실",<br/>"memo": "커트",<br/>"useDate": "2023-06-01",<br/>"payType": "현금" },<br/>{"recordId": 7,<br/>"amount": 1200,<br/>"category": "식비",<br/>"paidFor": "편의점",<br/>"memo": "껌",<br/>"useDate": "2023-06-01",<br/>"payType": "현금"}]| |
|성공 - 모든값 입력|memberId: 1<br/>startDate: 5/1/2023<br/>endDate: 6/30/2023<br/>categories: BEAUTY,CLOTHES|[{"recordId": 5,<br/>"amount": 20000,<br/>"category": "뷰티",<br/>"paidFor": "미용실",<br/>"memo": "커트",<br/>"useDate": "2023-06-01",<br/>"payType": "현금"},<br/>{"recordId": 4,<br/>"amount": 100000,<br/>"category": "의류",<br/>"paidFor": "백화점",<br/>"memo": "옷쇼핑",<br/>"useDate": "2023-05-05",<br/>"payType": "카드"}]|[{"recordId": 5,<br/>"amount": 20000,<br/>"category": "뷰티",<br/>"paidFor": "미용실",<br/>"memo": "커트",<br/>"useDate": "2023-06-01",<br/>"payType": "현금"},<br/>{"recordId": 4,<br/>"amount": 100000,<br/>"category": "의류",<br/>"paidFor": "백화점",<br/>"memo": "옷쇼핑",<br/>"useDate": "2023-05-05",<br/>"payType": "카드"}]| |
|실패 - 유효하지 않은 날짜|memberId: 1<br/>startDate: 66/1/2023<br/>endDate: 2/29/2024<br/>categories: FOOD|{"message": "유효하지 않은 날짜 형식입니다: 66/1/2023"}|{"message": "유효하지 않은 날짜 형식입니다: 66/1/2023"| |
|실패 - 유효하지 않은 날짜2|memberId: 1<br/>startDate: 2/29/2023<br/>endDate: 6/6/2023<br/>|{ "message": "유효하지 않은 날짜 형식입니다: 2/29/2023"}|{ "message": "유효하지 않은 날짜 형식입니다: 2/29/2023"}||
|실패 - 유효하지 않은 Category|startDate: 6/1/2023<br/>endDate: 2/29/2024<br/>categories: FOODS,BEAUTY|{"getInfos.categories": "Enum에 없는 값입니다"}|{"getInfos.categories": "Enum에 없는 값입니다"}| |
|실패 - 조회시작일이 조회종료일 이후|startDate: 6/1/2023<br/>endDate: 2/2/2020|콘솔창:<br/>java.lang.RuntimeException: 조회시작일이 조회종료일 이후입니다|콘솔창:<br/>java.lang.RuntimeException: 조회시작일이 조회종료일 이후입니다|Custom Exception과 연동 전으로 임시로 RuntimeException 발생<br/>Custom Exception으로 응답 수정 예정|

71mC40H0 committed:
test : 지출 등록, 수정, 삭제에 대한 테스트 코드 작성
- Login과 연동 시 테스트 코드 일부 수정 예정

71mC40H0 committed:
chore : querydsl 사용을 위한 querydsl 설정

71mC40H0 committed:
feat : List로 저장된 Enum 타입의 검증을 위한 custom annotation과 validator 구현

71mC40H0 committed:
feat : 프론트에서 제공한 LocalDate 양식에 맞는 converter 구현 및 등록

- converter 사용을 위한 config파일 설정 및 컨버터 등록
- [uuuu.M.d][uuuu.M.dd][uuuu.MM.d][uuuu.MM.dd] 모든 케이스에 대하여 적용
- ResolverStyle.STRICT 사용 이유: 해석하는 방식을 나타내는 ResolverStyle이 default의 경우 2023.2.30 같은 잘못된 경우를 자동으로 2023.2.28로 변경하기 때문
- yyyy가 아닌 uuuu 사용 이유:  ResolverStyle.STRICT에서는 yyyy 사용시 기원전/후를 표기해야 하기 때문

71mC40H0 committed:
refactor : 카테고리 변경에 따른 카테고리 수정

- 프론트의 요청에 맞게 수정

71mC40H0 committed:
refactor : 변수명 변경에 따른 변수명 수정 및 에러 메시지 수정

- 사용처: paidFor로 수정
- 에러 메세지: ExceptionHandler에 자동 생성된 에러메시지에 끝에 '.'이 없어 다른 메세지도 끝에 '.'삭제

71mC40H0 committed:
style : 코딩 스타일 통일에 따른 스타일 수정

- 맨 끝의 공백제거
- 맨 끝의 '}' 위에 공백 추가
- 맨 처음의 '{' 밑에 공백 추가

71mC40H0 committed:
feat : 가계부 메인화면 지출 내역 조회 기능 구현

- 조회는 기간과 카테고리를 통해 사용일 내림차순으로 조회
- 카테고리는 Optional
- 동적 쿼리 구현을 위한 querydsl 사용
- 응답을 위한 Response 클래스 구현

71mC40H0 committed:
test : 가계부 메인화면 지출 내역 조회 기능 테스트 코드 작성

71mC40H0 committed:
refactor : 가계부 메인화면 수정에 따른 Response 클래스 리팩토링

- payType 항목 추가

71mC40H0 committed:
test : RecordResponse 구조 변경에 따른 테스트 코드 수정

71mC40H0 committed:
refactor : form의 useDate 타입 수정

- 기존 deserializer로 처리할 경우, JsonParser에서 "2023.2.29" 같이 잘못 입력된 값을 "2023-02-28"로 변환하는 문제 발생
- 따라서 String으로 useDate를 받고 converter를 통해 변환하는 방식으로 수정

71mC40H0 committed:
refactor : 클래스명 및 패키지 변경, 응답방식 수정에 따른 코드 리팩토링

- Json형태로 응답을 위해 코드 수정

71mC40H0 committed:
feat : String에서 LocalDate 변환을 위한 util 클래스 구현

- 프론트에서 요청한 형식의 String을 LocalDate로 변환하기 위해 별도의 변환 방법이 필요
- 이미 작성한 Converter의 경우 사용하기 위해서는 객체 생성이 필요(webconfig 등록되어서 static 메소드 불가), 따라서 별도의 util 클래스 작성

71mC40H0 committed:
style : 붎필요한 줄 바꿈 삭제

71mC40H0 committed:
fix : ConstraintViolationException 처리를 위한 ExceptionHandler 추가

- API 테스트에서 ConstraintViolationException 발생 시 예상과 다른 결과 발생 확인
- ExceptionHandler 추가로 원하는 예상 결과 출력

NavyHubb commented on src/main/java/com/zerototen/savegame/service/RecordService.java:
```diff
@@ -1,51 +1,72 @@
+            throw new RuntimeException("Not match member"); // Login과 연동 시 CustomException으로 수정 예정
         }
 
         record.setAmount(serviceDto.getAmount());
```
serviceDto를 인자로 받아서 한번에 record 객체의 값을 수정하는 메서드를 만들면 좋을 것 같습니다.

71mC40H0 replied:
Record Entity에 serviceDto를 인자로 받아서 수정하는 메서드를 추가였습니다.

NavyHubb reviewed (approved):
리뷰 남겼습니다. 고생많으셨습니다!

71mC40H0 committed:
refactor : 코드 리뷰에 따른 코드 리팩토링

- update시 setter가 아닌 update 메서드를 통해 수정

71mC40H0 committed:
test : 코드 리뷰에 따른 테스트 코드 리팩토링

- given 메소드의 willReturn 부분을 한 줄 띄우는 형태로 스타일 수정
- given절의 객체 생성 부분을 별도의 private 메소드로 처리하는 방식으로 불필요한 반복 제거

71mC40H0 committed:
refactor : 날짜 양식 변경에 따른 날짜 패턴 변경 및 코드 리팩토링

- uuuu.M.d 패턴에서 M/d/uuuu 패턴으로 변경
- form의 useDate의 패턴에 @NotBlank가 포함되어 있으므로 어노테이션 삭제
- ExceptionHandler에 로그 추가

NavyHubb committed:
Merge branch 'develop' into feat-record-getInfos
